File: Leetcode/C++/543. Diameter of Binary Tree/solution.cpp
// https://leetcode.com/problems/diameter-of-binary-tree/

#include <algorithm>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int maxDist = 0;
    int diameterOfBinaryTree(TreeNode *root)
    {
        search(root);
        return maxDist;
    }

    int search(TreeNode *root)
    {
        if (root == NULL)
        {
            return 0;
        }
        int left = search(root->left), right = search(root->right);
        // Compare max distance and current path distance
        maxDist = max(maxDist, left + right);
        // Add current node to the longest distance.
        return max(left, right) + 1;
    }
};

File: Leetcode/Java/543. Diameter of Binary Tree/solution.java
// https://leetcode.com/problems/diameter-of-binary-tree/description/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    int longest = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        this.dfs(root);
        return this.longest;
    }

    public int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }

        int left = dfs(node.left);
        int right = dfs(node.right);
        // Check the diameter of the tree with the current node as the root.
        this.longest = Math.max(this.longest, left + right);

        return Math.max(left, right) + 1;
    }
}